### ISSUE TITLE
Ignore retracted versions of go dependencies

### ISSUE THREAD

**Issue by secustor** [renovatebot/renovate#13012]
### What would you like Renovate to be able to do?

Currently Renovate picks simply the tags of a dependency and uses semver to get the highest version.      
In case the highest version is retracted this this leads to unwanted upgrades.

 Example: 
https://pkg.go.dev/k8s.io/client-go@v1.5.2?tab=versions
https://github.com/kubernetes/client-go/blob/v1.5.2/go.mod

### If you have any ideas on how this should be implemented, please tell us here.

1. Get versions as it is done currently 
1. Fetch `go.mod` of highest semantic version
1. extract `retract` directive https://go.dev/ref/mod#go-mod-file-retract
1. Remove matching tags

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by rarkins** [comment]
Is this problem mostly avoided if GOPROXY is in use?

**Comment by secustor** [comment]
No, shouldn't make any difference as these proxies return all available versions even the retracted ones.     

Regardless, the implementation will be easier when using proxies as we can download the `go.mod` directly from the same API as the versions are queried.     
https://proxy.golang.org/k8s.io/client-go/@v/v1.5.2.mod

**Comment by sarahhodne** [comment]
As a workaround for the Kubernetes client-go library, I'm using this config to manually ignore those versions, in case anyone else is looking for something like this:

```json
{
  "packageRules": [
    {
      "matchPackageNames": ["k8s.io/client-go"],
      "allowedVersions": "!/1\\.(4\\.0|5\\.0|5\\.1|5\\.2)$/"
    }
  ]
}
```

**Comment by cevich** [comment]
I'm running into this problem as well now.  In case it helps, the version-retraction mechanism in go is explained here:

https://go.dev/ref/mod#go-mod-file-retract

It seems like the `rollback` feature should be in play here, but (clearly) isn't working for some reason.

**Comment by jsnod** [comment]
We just hit this, Renovate updated `github.com/prometheus/prometheus v1.8.2-0.20220202150209-4e08110891fd` to the retracted version: `github.com/prometheus/prometheus/v2 v2.35.0-retract`
